### ISSUE TITLE
NUMA awareness

### ISSUE THREAD

**Issue by okartau** [intel/pmem-csi#6]
NUMA-locality should be taken into account.

- when selecting region, the NUMA-local will provide the best performance.
  Q: NUMA-local compared to what? where driver itself runs?
  or should the API request carry some NUMA-related information?

- a namespace should not span to different NUMA nodes.
This gets covered by overall restriction that namespace should not span to different regions

**Comment by pohly** [comment]
Let's re-open, because there are new ideas around this. Whether anything needs to be done in PMEM-CSI remains to be seen.

The TopologyManager (https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0035-20190130-topology-manager.md) might take care of NUMA alignment, if kubelet can find the relevant information.

Several approaches are possible:
- somehow pass the information from PMEM-CSI to kubelet as part of the NodeStageVolume or NodePublishVolume responses (problem: nothing defined for this in the CSI spec)
- after PMEM-CSI has published a volume, kubelet could query the kernel and try to determine the CPU topology (problem: may be hard to do when abstractions like LVM are involved and/or interleaving gets added (issue #217))

/cc @kad

**Comment by okartau** [comment]
PMEM-CSI can find out associated numa_node for existing Regions and Namespaces, using libndctl API:
```
libndctl.h:int ndctl_region_get_numa_node(struct ndctl_region *region);
libndctl.h:int ndctl_namespace_get_numa_node(struct ndctl_namespace *ndns);
```
Currently we do not retrieve or use such information. As stated, CSI spec has no method passing it.

**Comment by pohly** [comment]
> PMEM-CSI can find out associated numa_node for existing Regions and Namespaces, using libndctl API:
> 
> libndctl.h:int ndctl_region_get_numa_node(struct ndctl_region *region);
> libndctl.h:int ndctl_namespace_get_numa_node(struct ndctl_namespace *ndns);

Does that also give reasonable results when running in a VM?

**Comment by okartau** [comment]
> Does that also give reasonable results when running in a VM?

Yes: It seems to show similar results as on bare metal, but to see numa_node reported, you have to run on a qemu guest with NUMA nodes configured (and emulated nvdimm devices associated with those emulated NUMA nodes).

I tried such config with 2 NUMA nodes and 2 emulated NVDIMMs, one per node:
`ndctl list -Rv` output (-v option enables more detailed part where numa_node is reported)
 is similar to what I see on bare metal: numa_node is reported for Regions and Namespaces.

On a qemu guest without NUMA nodes, the numa_node part is not reported by ndctl. (this is what we use now in test setup).

**Comment by liverbirdkte** [comment]
Hi guys, any updates on this NUMA awareness issue? 
I'm working on some benchmarks to evaluating performance of AD mode pmem Redis in Kubernetes,  is there a way to guarantee that service pod and pmem are in the same NUMA node?

**Comment by avalluri** [comment]
> Hi guys, any updates on this NUMA awareness issue?
> I'm working on some benchmarks to evaluating performance of AD mode pmem Redis in Kubernetes, is there a way to guarantee that service pod and pmem are in the same NUMA node?

As of now, there is no update yet on this issue, It's still under discussion how the driver provide and consume(in late binding case) numa information.

**Comment by pohly** [comment]
@avalluri: it looks like your worked on this (https://github.com/intel/pmem-csi/pull/392#issuecomment-525205832). Can you summarize the current status and who's doing what next?

**Comment by avalluri** [comment]
> @avalluri: it looks like your worked on this (#392 (comment)). Can you summarize the current status and who's doing what next?

No really much, except the #392 which adds numa node awareness to driver - holding numa information about created namespace device,, but does nothing special with that information currently.

As per @kad's praposal,  the [CRI Resource Manager](https://github.com/intel/cri-resource-manager)  magically places the Pods/containers to the right CPUs by detecting the devices/volumes its using, in case of a PMEM device it considers the numa node of the attached device. But still trying the PoC with this concept is pending.

**Comment by pohly** [comment]
@avallari: if the CRI Resource Manager is detecting this itself, then we don't need code for it in the PMEM-CSI driver.

@kad: are their tracking issues for whatever is missing in CRI Resource Manager? Who is driving this?

**Comment by okartau** [comment]
@pohly you added 0.7 label, but is there anything that can be done about this topic in pmem-csi scope for 0.7 release?
Note that related #392 which is about adding Numa-related code, has been stalled for 6 months (is out of sync) only adds code "for debug purpose", so it does not add any functional value to the story.
Overall, CSI spec still does not have anything that would help NUMA topic, all relevant decisions (like scheduling) happen elsewhere, so seems we can't make any improving changes in pmem-csi scope. What about closing this issue (it did not have clear focus from very start). In the future, if the need would rise, we can create another more focused topic.

**Comment by pohly** [comment]
@kad pointed out on the internal chat that this gets solved by CRI-R-M when using a "the topology aware policy with memory tiering support" as explained in https://github.com/intel/cri-resource-manager/blob/b27d2295849ae9ca6e5e8d3fb6b2f03ff7543f83/README.md#tldr-setup

We could document this in our repo, but it's simpler to just point people towards CRI-R-M when the question comes up again. Therefore we can close the issue.
